File: src/patterns/behavioral/observer/observer.ts
// Publisher
class ChatRoom {
  users: IUser[] = [];
  messages: string[] = [];
  constructor(users: IUser[], messages?: string[]) {
    this.users = users;
    this.messages = messages ? messages : [];
    for (let i = 0; i < users.length; i++) {
      this.users[i].setChatRoom(this);
    }
  }

  addUser(user: IUser): void {
    if (!this.users.includes(user)) {
      this.users.push(user);
      user.setChatRoom(this);
    }
  }

  removeUser(user: IUser): void {
    if (this.users.includes(user)) {
      this.users = this.users.filter((u) => u !== user);
    }
  }

  notifyUsers(message: string, sentBy: IUser) {
    this.messages.push(message);
    const usersToNotify = this.users.filter((u) => u !== sentBy);
    for (let i = 0; i < usersToNotify.length; i++) {
      usersToNotify[i].update(message);
    }
  }
}

// Subscribers
interface IUser {
  sendMessage(message: string): void;
  update(message: string): void;
  setChatRoom(chatRoom: ChatRoom): void;
}

class User implements IUser {
  chatRoom: ChatRoom | null;
  name: string;
  age: number;
  constructor(name: string, age: number, chatRoom?: ChatRoom) {
    this.chatRoom = chatRoom ? chatRoom : null;
    this.name = name;
    this.age = age;
  }

  setChatRoom(chatRoom: ChatRoom) {
    this.chatRoom = chatRoom;
  }

  sendMessage(message: string): void {
    this.chatRoom?.notifyUsers(message, this);
  }

  update(message: string): void {
    console.log("Message received by user " + this.name + ": " + message);
  }
}

const user1 = new User("Nikita", 20);
const user2 = new User("Olya", 20);
const user3 = new User("Sasha", 10);
const user4 = new User("Bebra", 3);

const chatRoom = new ChatRoom([user1, user2, user3]);

chatRoom.addUser(user4);

user1.sendMessage("hey");
console.log(chatRoom.messages);

File: src/patterns/behavioral/observer/observer.go
package main

import "fmt"

// Publisher
type ChatRoom struct {
	users []IUser
	messages []string
}

func NewChatRoom(users []IUser) *ChatRoom {
	chatRoom := &ChatRoom{users: users}

	for _, u := range users {
		u.SetChatRoom(chatRoom)
	}

	return chatRoom
}

func (r *ChatRoom) addUser(user IUser) {
	isAdded := false
	for _, u := range r.users {
		if u == user {
			isAdded = true
		}
	}

	if !isAdded {
		r.users = append(r.users, user)
		user.SetChatRoom(r)
	}
}

func (r *ChatRoom) removeUser(user IUser) {
	filteredUsersArr := make([]IUser, 0, len(r.users)-1)
	for _, u := range r.users {
		if u != user {
			filteredUsersArr = append(filteredUsersArr, u)
		}
	}

	r.users = filteredUsersArr
}

func (r *ChatRoom) NotifyUsers(message string, sentBy IUser){
	r.messages = append(r.messages, message)
	for _, u := range r.users {
		if u != sentBy {
			u.Update(message)
		}
	}
}

func (r *ChatRoom) MakeAnnouncement(message string){
	FibonacciCalc(10)
	for _, u := range r.users {
		u.Update(message)
	}
}

// Subscribers
type IUser interface {
	Update(message string)
	SendMessage(message string)
	SetChatRoom(store *ChatRoom)
}

type User struct {
	name string
	chatRoom *ChatRoom
}

func (u *User) SetChatRoom(room *ChatRoom){
	u.chatRoom = room
}

func (u *User) SendMessage(message string){
	u.chatRoom.NotifyUsers(message, u)
}

func (u *User) Update(message string){
	fmt.Printf("Message received by user %v: %v \n", u.name, message)
}

// Client

// Fibonacci Sequence
func FibonacciCalc(num int) int {
	if num <= 1 {
		return num
	}

	return FibonacciCalc(num-1) + FibonacciCalc(num-2)
}

// func main(){
// 	user1 := &User{name: "user1"}
// 	user2 := &User{name:"user2"}
// 	user3 := &User{name:"user3"}
// 	user4 := &User{name:"user4"}

// 	chatRoom := NewChatRoom([]IUser{user1, user2, user3})
// 	chatRoom.addUser(user4)

// 	user1.SendMessage("hey")

// 	fmt.Println(chatRoom.messages)

// 	chatRoom.MakeAnnouncement("I HAVE GOT AN ANNOUNCEMENT TO MAKE")
// }
